File: cs25backend/src/main/java/edu/dongguk/cs25backend/domain/Manager.java
package edu.dongguk.cs25backend.domain;

import edu.dongguk.cs25backend.domain.type.MemberShip;
import edu.dongguk.cs25backend.domain.type.UserRole;
import jakarta.persistence.*;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.DynamicUpdate;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@NoArgsConstructor
@DynamicUpdate
@Getter
@Table(name = "managers")
public class Manager {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "manager_id")
    private Long id;

    @JoinColumn(name = "login_id", nullable = false)
    private String loginId;

    @JoinColumn(name = "password", nullable = false)
    private String password;

    @JoinColumn(name = "name", nullable = false)
    private String name;

    @JoinColumn(name = "email", nullable = false)
    private String email;

    @JoinColumn(name = "phone_number", nullable = false)
    private String phoneNumber;

    @Enumerated(EnumType.STRING)
    @JoinColumn(name = "user_role", nullable = false)
    private UserRole userRole;

    @Enumerated(EnumType.STRING)
    @JoinColumn(name = "membership", nullable = false)
    private MemberShip memberShip;

    @Column(updatable = false)
    @JoinColumn(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @JoinColumn(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    //== 연관 관계 매핑 ==//
    @OneToMany(mappedBy = "manager")
    private List<Store> stores = new ArrayList<>();

    @Builder
    public Manager(String loginId, String password, String name, String email, String phoneNumber, UserRole userRole, MemberShip memberShip) {
        this.loginId = loginId;
        this.password = password;
        this.name = name;
        this.email = email;
        this.phoneNumber = phoneNumber;
        this.userRole = userRole;
        this.memberShip = memberShip;
        this.createdAt = LocalDateTime.now();
    }
}

File: cs25server/src/main/kotlin/edu/dongguk/cs25server/domain/Manager.kt
package edu.dongguk.cs25server.domain

import edu.dongguk.cs25server.domain.type.AllowStatus
import edu.dongguk.cs25server.domain.type.Membership
import edu.dongguk.cs25server.domain.type.UserRole
import jakarta.persistence.*
import org.hibernate.annotations.DynamicUpdate
import java.time.LocalDate
import java.time.LocalDateTime

@Entity
@DynamicUpdate
@Table(name = "managers")
class Manager(
    @Column(name = "login_id", nullable = false)
    val loginId: String,

    @Column(name = "password", nullable = false)
    val password: String,

    @Column(name = "name", nullable = false)
    val name: String,

    @Column(name = "email", nullable = false)
    val email: String,

    @Column(name = "phone_number", nullable = false)
    val phoneNumber: String,

    @Enumerated(EnumType.STRING)
    @Column(name = "role", nullable = false)
    val role: UserRole = UserRole.MANAGER,

    @Enumerated(EnumType.STRING)
    @Column(name = "membership", nullable = false)
    val memberShip: Membership = Membership.NORMAL,

    @Column(name = "created_at", nullable = false, updatable = false)
    val createdAt: LocalDate = LocalDate.now(),

    @Column(name = "status", nullable = false)
    @Enumerated(EnumType.STRING)
    var status: AllowStatus = AllowStatus.BEFORE
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "manager_id")
    private val id: Long? = null

    @Column(name = "is_login")
    private var isLogin: Boolean? = null

    @Column(name = "refresh_token")
    private var refreshToken: String? = null

    @JoinColumn(name = "updated_at", nullable = false)
    private val updatedAt: LocalDateTime? = null

    @OneToMany(mappedBy = "manager")
    private val stores: List<Store> = ArrayList()

    fun getId() = this.id

    fun setLogin(refreshToken: String) {
        this.refreshToken = refreshToken
        this.isLogin = true
    }

    fun setLogout() {
        this.refreshToken = refreshToken
        this.isLogin = false
    }

    fun allowManager(status: AllowStatus) {
        this.status = status;
    }

}


